File: jrand-annotation-processor/src/main/java/me/xdrop/jrand/annotation/processing/FacadeProcessor.java
package me.xdrop.jrand.annotation.processing;

import com.github.javaparser.ast.CompilationUnit;
import com.squareup.javapoet.TypeSpec;
import me.xdrop.jrand.annotation.Facade;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.PackageElement;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import java.io.*;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;


@SupportedAnnotationTypes("me.xdrop.jrand.annotation.Facade")
@SupportedSourceVersion(SourceVersion.RELEASE_7)
public class FacadeProcessor extends BaseProcessor {
    private FacadeClassBuilder classBuilder;
    private ForkClassGenerator forkClassGenerator;
    private Map<String, TypeElement> facadeClasses;
    private Path outputPathFacade;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        this.facadeClasses = new HashMap<>();
        this.classBuilder = new FacadeClassBuilder();
        this.forkClassGenerator = new ForkClassGenerator(processingEnv, getRepository());
        this.outputPathFacade = Paths.get("jrand-core", "src", "generated", "java", "me", "xdrop", "jrand");
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        for (Element element : roundEnv.getElementsAnnotatedWith(Facade.class)) {
            if (element.getKind() != ElementKind.CLASS) {
                getMessager().printMessage(Diagnostic.Kind.ERROR, "Annotation [Facade] should only be used on classes.");
                return true;
            }

            TypeElement typeElement = (TypeElement) element;
            PackageElement pkg = processingEnv.getElementUtils().getPackageOf(typeElement);

            Facade facade = element.getAnnotation(Facade.class);
            String accessor = facade.accessor();
            facadeClasses.put(accessor, typeElement);

            try {
                // Get the last part of the package eg. me.xdrop.generators.text would be text
                String lastPackage = getLastPackageName(pkg);
                // Get the class name eg. LoremGenerator
                String className = typeElement.getSimpleName().toString();
                // Get the output path which eg. generated/me/xdrop/generators/text
                Path path = Paths.get(getOutputPathGenerators().toString(), lastPackage);
                boolean mkdirs = path.toFile().mkdirs();
                // Get the source Compilation unit
                CompilationUnit sourceCU = getRepository().getCU(pkg.toString(), className);
                // Create the new compilation unit
                CompilationUnit newCU = forkClassGenerator.buildForkedClass(typeElement, className, sourceCU);
                getRepository().addCU(className, newCU);
                // Write the new compilation unit
                getRepository().writeTo(path, className, newCU);
            } catch (IOException e) {
                e.printStackTrace();
            }

        }

        TypeSpec jrandFacade = classBuilder.buildFacadeClass(facadeClasses);

        try {
            try (FileWriter fw = new FileWriter(new File(outputPathFacade.toString(), "JRand.java"))) {
                classBuilder.writeFacade(fw, jrandFacade, packageName);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;

    }



    private class FacadeGeneratorInfo {
        private String packageName;
        private String simpleName;
    }
}

File: jrand-annotation-processor/src/main/java/me/xdrop/jrand/annotation/processing/FacadeProcessor.kt
package me.xdrop.jrand.annotation.processing

import com.squareup.javapoet.JavaFile
import me.xdrop.jrand.annotation.Facade
import me.xdrop.jrand.annotation.PropertyFlag
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.ElementKind
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic


/**
 * The facade processor is responsible for creating the JRand facade class,
 * as well as rendering the finalized extended generators with the
 * fork() method as well as any existing modifications made from
 * the property processor.
 */
@SupportedAnnotationTypes("me.xdrop.jrand.annotation.Facade")
@SupportedSourceVersion(SourceVersion.RELEASE_7)
class FacadeProcessor : BaseProcessor() {
    private val classBuilder: FacadeClassBuilder = FacadeClassBuilder()
    private var forkClassGenerator: ForkClassGenerator? = null
    private val facadeClasses: MutableMap<String, Pair<TypeElement, String>> = HashMap()

    @Synchronized
    override fun init(processingEnv: ProcessingEnvironment) {
        super.init(processingEnv)
        this.forkClassGenerator = ForkClassGenerator(processingEnv)
    }

    override fun process(annotations: Set<TypeElement>, roundEnv: RoundEnvironment): Boolean {
        if (roundEnv.processingOver()) {
            return true
        }

        roundEnv.getElementsAnnotatedWith(Facade::class.java).forEach { element ->
            if (element.kind != ElementKind.CLASS || element !is TypeElement) {
                messager.printMessage(Diagnostic.Kind.ERROR, "Annotation [Facade] should only be used on classes.")
                return true
            }

            val pkg = processingEnv.elementUtils.getPackageOf(element)
            val annotation = element.getAnnotation(Facade::class.java)
            val id = ClassIdentifier(element, pkg)
            val accessor = annotation.accessor

            ProcessorRepository.addMethods(id, forkClassGenerator?.getForkAndCloneMethods(element) ?: emptyList())
            ProcessorRepository.writeToFiler(id, filer)

            facadeClasses[accessor] = Pair(element, pkg.toString())
        }

        val jrandFacade = classBuilder.buildFacadeClass(facadeClasses)

        try {
            JavaFile.builder("me.xdrop.jrand", jrandFacade)
                    .build().writeTo(filer)
        } catch (ignored: FilerException) {
            // This fails because on later rounds the facade is already created, but it is fine
        }

        return false
    }

}
